Closes #2324 - Benchmark documentation (#2361)
Bears-R-Us/arkouda: pydoc/developer/BENCHMARK.md

Author: Done

# PyTest Benchmarks

**Purpose**: Outline the general steps and options available for running the new PyTest benchmarks

The PyTest benchmark structure allows the flexibility of running the entire suite, a single file, 
or a single benchmark.

## Running The Full Suite
In most cases, running the full benchmark suite is desired. The simplest way to do this is to navigate into the 
root-level of arkouda and run `make benchmark`

This will run the entire benchmark suite with the following command:
```commandline
python3 -m pytest -c benchmark.ini --benchmark-autosave --benchmark-storage=file://benchmark_v2/.benchmarks

> ```

Reviewer: After this block, we should mention that the resulting JSON files will be located in the `arkouda/benchmarks_v2/.benchmarks` folder. It may be useful to identify how the files are named as well.

Author: Done

## Running Single Files or Tests

> In instances where a single test or set of tests needs to be ran, use the `-k <name_of_file>.py` argument. For these 
> examples, `encoding_benchmark.py` and the specific benchmark from within the file `bench_encode` is used.

Reviewer: ```suggestion
In instances where a single test or set of tests needs to be ran, use the `-k <name_of_file>.py` argument. 
```

Following the above, I would add the information the first example will run all the benchmarks in the `encoding_benchmark.py` file.

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
bench_encode[idna]      3.3304 (1.0)      9.2561 (2.10)     4.7544 (1.27)     2.5306 (6.18)     3.8075 (1.11)     1.9012 (3.62)          1;1  210.3306 (0.79)          5           1
bench_encode[ascii]     3.3805 (1.02)     4.8800 (1.10)     3.7336 (1.0)      0.6465 (1.58)     3.4231 (1.0)      0.5246 (1.0)           1;1  267.8380 (1.0)           5           1
bench_decode[idna]      3.4444 (1.03)     4.4177 (1.0)      3.7852 (1.01)     0.4097 (1.0)      3.5622 (1.04)     0.5837 (1.11)          1;0  264.1882 (0.99)          5           1
bench_decode[ascii]     3.4621 (1.04)     4.9177 (1.11)     4.2250 (1.13)     0.6125 (1.50)     4.0197 (1.17)     0.9991 (1.90)          2;0  236.6864 (0.88)          5           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

> Similarly, to only run a single test within a file, specify the test name with the `-k` flag instead of a filename

Reviewer: ```suggestion
Similarly, to only run a single test within a file, specify the test name with the `-k` flag instead of a filename. The following example will run only the `bench_encode` benchmark.
```

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
bench_encode[ascii]     3.4298 (1.0)      3.6450 (1.0)      3.5541 (1.0)      0.0889 (1.0)      3.5801 (1.00)     0.1436 (1.0)           2;0  281.3620 (1.0)           5           1
bench_encode[idna]      3.4875 (1.02)     4.5255 (1.24)     3.7912 (1.07)     0.4328 (4.87)     3.5652 (1.0)      0.4869 (3.39)          1;0  263.7659 (0.94)          5           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

## Benchmark Arguments
There are a large number of commandline arguments available for configuring the benchmarks to run in a way fitting 
to any use case.

<table>
<tr><th>Argument</th><th>Default Value</th></tr>
<tr><th colspan="2">Description</th></tr>
<tr>
    <td>--benchmark-autosave</td>
    <td>True when using "<b>make benchmark"</b></td>
</tr>

> <tr><td colspan="2">Save the benchmark JSON results to the provided storage location</td></tr>

Reviewer: Instead of using a table here I would do something like 

```
`--benchmark-autosave`
> Used by default when running `make benchmark`.
> Save the benchmark JSON results to the provided storage location
```
Then for parameters that truly have a default do something like
```
`--size`
> **Default:** 10**8
> Problem size: length of array to use for benchmarks.
```

Author: Added

<tr>
    <td>--benchmark-storage</td>
    <td>"file://benchmark_v2/.benchmarks" when using "<b>make benchmark"</b></td>
</tr>

> <tr><td colspan="2">Storage location to save benchmark output JSON</td></tr>

Reviewer: ```suggestion
<tr><td colspan="2">Storage location for benchmark output JSON</td></tr>

```

<tr>
    <td>--benchmark-save</td>
    <td>"COMMIT-ID_DATE_TIME_IS-DIRTY" example: 

> 31de39be8b19c76d073a8999def6673a305c250d_20230405_145947_uncommited-changes.json</td>

Reviewer: Probably want one that is clean here because most users won't have this occur.

Author: Updated

</tr>
<tr><td colspan="2">Name to save the output JSON as. Will be saved as "counter_NAME.json"</td></tr>
<tr>
    <td>--size</td>
    <td>10**8</td>
</tr>
<tr><td colspan="2">Problem size: length of array to use for benchmarks.</td></tr>
<tr>
    <td>--trials</td>
    <td>5</td>
</tr>
<tr><td colspan="2">Number of times to run each test before averaging results. For tests that run as many trials as possible in 
a given time, will be treated as number of seconds to run for.</td></tr>
<tr>
    <td>--seed</td>
    <td></td>
</tr>
<tr><td colspan="2">Value to initialize random number generator.</td></tr>
<tr>
    <td>--dtype</td>
    <td></td>
</tr>
<tr><td colspan="2">Dtypes to run benchmarks against. Comma separated list (NO SPACES) allowing for multiple. Accepted values: 

> int64, uint64, bigint, float64, bool, str and mixed. Mixed is used to generate sets of multiple types.</td></tr>

Reviewer: Can we provide an example here of what the usage looks like?

<tr>
    <td>--numpy</td>
    <td>False</td>
</tr>

> <tr><td colspan="2">When set, runs numpy comparison benchmarks.</td></tr>

Reviewer: Mark this as a `stores true` so users know they do not need to provide the value.

Author: Added

<tr>
    <td>--maxbits</td>
    <td>-1</td>
</tr>

> <tr><td colspan="2">Only applies to bigint testing. Maximum number of bits, so values > 2**max_bits will wraparound. -1 is 

Reviewer: ```suggestion
<tr><td colspan="2">Only applies to bigint. Maximum number of bits, so values > 2**max_bits will wraparound. -1 is 

```

interpreted as no maximum.</td></tr>
<tr>
    <td>--alpha</td>
    <td>1.0</td>
</tr>
<tr><td colspan="2">Scalar multiple</td></tr>
<tr>
    <td>--randomize</td>
    <td>False</td>
</tr>
<tr><td colspan="2">Fill arrays with random values instead of ones</td></tr>
<tr>
    <td>--index_size</td>
    <td></td>
</tr>
<tr><td colspan="2">Only used for Gather and Scatter Benchmarks. Length of index array (number of gathers to perform)</td></tr>
<tr>
    <td>--value_size</td>
    <td></td>
</tr>
<tr><td colspan="2">Only used for Gather and Scatter Benchmarks. Length of array from which values are gathered</td></tr>
<tr>
    <td>--encoding</td>
    <td></td>
</tr>
<tr><td colspan="2">Only applies to encoding benchmarks. Comma separated list (NO SPACES) allowing for multiple encoding to be 

> used. Accepted values: idna, ascii</td></tr>

Reviewer: Can we provide a example?

Author: Added

<tr>
    <td>--io_only_write</td>
    <td>False</td>
</tr>
<tr><td colspan="2">Only applies to IO benchmarks. Only write the files; files will not be removed</td></tr>
<tr>
    <td>--io_only_read</td>
    <td>False</td>
</tr>
<tr><td colspan="2">Only applies to IO benchmarks. Only read the files; files will not be removed</td></tr>
<tr>
    <td>--io_only_delete</td>
    <td>False</td>
</tr>
<tr><td colspan="2">Only applies to IO benchmarks. Only delete files created from writing with this benchmark</td></tr>
<tr>
    <td>--io_files_per_loc</td>
    <td>1</td>
</tr>
<tr><td colspan="2">Only applies to IO benchmarks. Number of files to create per locale</td></tr>
<tr>
    <td>--io_compression</td>
    <td></td>
</tr>
<tr><td colspan="2">Only applies to IO benchmarks. Compression types to run Parquet IO benchmarks against. Comma delimited list 

> (NO SPACES) allowing for multiple. Accepted values: none, snappy, gzip, brotli, zstd, and lz4</td></tr>

Reviewer: Can we provide an example of this being used?

Author: Added

    "benchmarks": [
        {
            "group": "Strings_EncodeDecode",
            "name": "bench_encode[ascii]",
            "params": {
                "encoding": "ascii"
            },
            "extra_info": {
                "description": "Measures the performance of Strings.encode",
                "problem_size": 100,
                "transfer_rate": "0.0002 GiB/sec"
            },
            "stats": {
                "min": 0.00383609999971668,
                "max": 0.0043372999998609885,
                "mean": 0.004057779999857303,
                "stddev": 0.00018361238254747651,
                "rounds": 5,
                "median": 0.0040258999997604406,
                "iqr": 0.0002090000002681336,
                "q1": 0.0039507749997937935,
                "q3": 0.004159775000061927,
                "iqr_outliers": 0,
                "stddev_outliers": 2,
                "outliers": "2;0",
                "ld15iqr": 0.00383609999971668,
                "hd15iqr": 0.0043372999998609885,
                "ops": 246.44017172817806,
                "total": 0.020288899999286514
            }
        }
    ],
    "datetime": "2023-04-05T15:32:09.097392"
}
```

> The main important components are `benchmarks.extra_info`, which contains the details on the problem size and 

Reviewer: ```suggestion
The components to pay attention to are `benchmarks.extra_info`, which contains the details on the problem size and 

```

# PyTest Benchmarks

Arkouda uses `pytest-benchmark` for performance benchmarking. This document provides an overview of running `pytest-benchmark` and the configurations available to the user.

More information on `pytest-benchmark` can be found [here](https://pytest-benchmark.readthedocs.io/en/latest/)

## Running The Full Suite
In most cases, running the full benchmark suite is desired. The simplest way to do this is to navigate into the 
root-level of arkouda and run `make benchmark`

This will run the entire benchmark suite with the following command:
```commandline
python3 -m pytest -c benchmark.ini --benchmark-autosave --benchmark-storage=file://benchmark_v2/.benchmarks
```

> The results for the benchmarks can be found within the provided benchmark storage path, which by default is within a
> directory found in `//benchmark_v2/.benchmarks`.
> 
> The command used here will provide a full set of information on all benchmarks available. 

Reviewer: ```suggestion
The results for the benchmarks can be found within the provided benchmark storage path, which by default is within a directory found in `//benchmark_v2/.benchmarks`. Here you will find JSON files with the details on all the benchmarks.
```

Author: Merged Suggestion

The `-c` flag specifies to PyTest to use `benchmark.ini` as the configuration file for this set of test. The 
configuration file specifies which files contain benchmarks as well as a set of environment variables used by 
the benchmarks.

`--benchmark-autosave` tells `pytest-benchmark` to save the results of the benchmark in a json file stored in the

> path specified by `--benchmark-storage`.

Reviewer: I would leave these here, but also I was hoping to add `-c` and `-k` to the `Benchmark Arguments` Section.

Author: Added -c and -k to the arguments section

For more information on these and other available flags, look in the [Benchmark Arguments](#args) section of this
document.

<span id="args"></span>
## Benchmark Arguments
There are a large number of commandline arguments available for configuring the benchmarks to run in a way fitting 
to any use case.

`--benchmark-autosave`
> Used by default when running `make benchmark`
> 
> Save the benchmark JSON results to the provided storage location

`--benchmark-storage`
> Sets location to "file://benchmark_v2/.benchmarks" when using `make benchmark`
> 
> Storage location for benchmark output JSON

`--benchmark-save`

> > **Default:** "COMMIT-ID_DATE_TIME_IS-DIRTY" example: 0d4865d7c9453adc6af6409568da326845c358b9_20230406_165330.json

Reviewer: ```suggestion
> example: 0d4865d7c9453adc6af6409568da326845c358b9_20230406_165330.json

```
